### DISCUSSION TITLE
TRIM support in BurmillaOS

### DISCUSSION THREAD

**Discussion by MattKobayashi** [burmilla/os#135] [Q&A] [answered]
Hi all, I'm trying to understand a recent significant decrease in I/O performance on my BurmillaOS installation, one of the avenues I'm exploring is ensuring that the SSD that the OS is installed on (baremetal install) has been TRIMed.

From my investigations I've noted that /dev/sda1 is mounted with the following options:

`rw,relatime,data=ordered`

The `discard` option isn't enabled so periodic TRIM operations wouldn't seem to be happening. `fstrim` isn't installed on the OS by default, and I'm not sure of a way to get it on the system (running it from a container with /dev/sda or /dev/sda1 mounted did not work).

Wondering if anyone has any experience with this that they could share?

**Comment by MattKobayashi** [comment, marked as answer]
Well I'm silly. Turns out `fstrim` does exist in BurmillaOS, just needs to be run as `sudo`, e.g. `sudo fstrim -av`. Posting this here in case anyone else has a similar question in future.

**Reply by olljanat** [reply]
> The discard option isn't enabled so periodic TRIM operations wouldn't seem to be happening.

Hmm. I not familiar with this area but so if we would have `discard` on mount parameters it would simplify situation for you?

We have added some of the optimizations recommend by Nutanix on https://portal.nutanix.com/page/documents/solutions/details/?targetId=BP-2105-Linux-on-AHV:BP-2105-Linux-on-AHV and looks that they would parameters `noatime,nodiratime,discard,barrier=0` so I think that we can consider using those on next version.

**Reply by MattKobayashi** [reply]
Hi @olljanat, having the `discard` option enabled on the mount is definitely an idea worth considering. There are some [varying opinions](https://www.digitalocean.com/community/tutorials/how-to-configure-periodic-trim-for-ssd-storage-on-linux-servers) on the use of continuous TRIM due to additional overhead on normal file operations, but I'm not sure that the performance hit would be enough to negate it as an option.

At the moment I've implemented the `fstrim` command as a runcmd in cloud-config, so it will at least run on every boot of the server. Perhaps another option might be to set up a default system cron task to run `fstrim` on a regular (perhaps monthly?) basis.
